File: Balance.kt
import java.io.File
import java.util.*

object Balance {

    private const val FILE_VERSION = 1
    private const val MAGIC_CONST = 0x5F3759DF

    @JvmStatic
    fun main() {
        print("请输入货币数量: ")
        val input = readlnOrNull()?.toLongOrNull()

        if (input == null || input < 0 || input > Int.MAX_VALUE) {
            println("输入无效，必须是 0 ~ ${Int.MAX_VALUE}")
            return
        }

        val value = input.toInt()
        val bytes = generateBalanceFile(value)

        println("\n生成的 balance.dat 十六进制数据：")
        println(bytes.joinToString(" ") { "%02X".format(it) })

        File("balance.dat").writeBytes(bytes)

        // 额外校验
        println("\n校验解码余额: ${decodeBalance(bytes)}")
    }

    fun generateBalanceFile(value: Int): ByteArray {
        val rng = Random()
        var xorKey = rng.nextInt()
        if (xorKey == 0) {
            xorKey = 1597463007
        }

        val obfuscated = value xor xorKey
        val checksum = computeChecksum(value, xorKey)
        val signature = computeFileSignature(FILE_VERSION, xorKey, obfuscated, checksum)

        val buffer = ByteArray(20)

        writeIntLE(buffer, 0, FILE_VERSION)
        writeIntLE(buffer, 4, xorKey)
        writeIntLE(buffer, 8, obfuscated)
        writeIntLE(buffer, 12, checksum)
        writeIntLE(buffer, 16, signature)

        return buffer
    }

    private fun computeChecksum(value: Int, key: Int): Int {
        var num = value
        num = (num * 397) xor key
        num = num xor MAGIC_CONST
        return rotl(num, 7)
    }

    private fun computeFileSignature(
        version: Int,
        xorKey: Int,
        obfuscated: Int,
        checksum: Int
    ): Int {
        var num = version
        num = (num * 31) xor xorKey
        num = (num * 31) xor obfuscated
        num = (num * 31) xor checksum
        num = num xor MAGIC_CONST
        return rotl(num, 11)
    }

    private fun rotl(value: Int, bits: Int): Int {
        return (value shl bits) or (value shr (32 - bits))
    }

    private fun writeIntLE(buffer: ByteArray, offset: Int, value: Int) {
        buffer[offset] = (value and 0xFF).toByte()
        buffer[offset + 1] = ((value shr 8) and 0xFF).toByte()
        buffer[offset + 2] = ((value shr 16) and 0xFF).toByte()
        buffer[offset + 3] = ((value shr 24) and 0xFF).toByte()
    }

    private fun readIntLE(buffer: ByteArray, offset: Int): Int {
        return (buffer[offset].toInt() and 0xFF) or
                ((buffer[offset + 1].toInt() and 0xFF) shl 8) or
                ((buffer[offset + 2].toInt() and 0xFF) shl 16) or
                ((buffer[offset + 3].toInt() and 0xFF) shl 24)
    }

    // 用于测试生成结果是否正确
    fun decodeBalance(buffer: ByteArray): Int {
        val xorKey = readIntLE(buffer, 4)
        val obfuscated = readIntLE(buffer, 8)
        return obfuscated xor xorKey
    }
}

File: balance.cpp
#include <iostream>
#include <fstream>
#include <cstdint>
#include <random>

using namespace std;

static const int32_t FILE_VERSION = 1;
static const int32_t MAGIC_CONST = 0x5F3759DF;

static inline int32_t rotl(int32_t value, int bits)
{
	return (value << bits) | (value >> (32 - bits));
}

static int32_t computeChecksum(int32_t value, int32_t key)
{
	int32_t num = value;
	num = (num * 397) ^ key;
	num = num ^ MAGIC_CONST;
	return rotl(num, 7);
}

static int32_t computeFileSignature(
									int32_t version,
									int32_t xorKey,
									int32_t obfuscated,
									int32_t checksum)
{
	int32_t num = version;
	num = (num * 31) ^ xorKey;
	num = (num * 31) ^ obfuscated;
	num = (num * 31) ^ checksum;
	num = num ^ MAGIC_CONST;
	return rotl(num, 11);
}

static void writeIntLE(ofstream& file, int32_t value)
{
	file.put(static_cast<char>(value & 0xFF));
	file.put(static_cast<char>((value >> 8) & 0xFF));
	file.put(static_cast<char>((value >> 16) & 0xFF));
	file.put(static_cast<char>((value >> 24) & 0xFF));
}

int main(int argc, char* argv[])
{
	if (argc < 2)
	{
		cout << "Usage: balance.exe <money>\n";
		return 0;
	}
	
	int64_t input = stoll(argv[1]);
	if (input < 0 || input > INT32_MAX)
	{
		cout << "Money must be between 0 and " << INT32_MAX << endl;
		return 0;
	}
	
	int32_t value = static_cast<int32_t>(input);
	
	random_device rd;
	mt19937 gen(rd());
	uniform_int_distribution<int32_t> dist(INT32_MIN, INT32_MAX);
	
	int32_t xorKey = dist(gen);
	if (xorKey == 0)
		xorKey = 1597463007;
	
	int32_t obfuscated = value ^ xorKey;
	int32_t checksum = computeChecksum(value, xorKey);
	int32_t signature =
	computeFileSignature(FILE_VERSION, xorKey, obfuscated, checksum);
	
	ofstream file("balance.dat", ios::binary | ios::trunc);
	
	writeIntLE(file, FILE_VERSION);
	writeIntLE(file, xorKey);
	writeIntLE(file, obfuscated);
	writeIntLE(file, checksum);
	writeIntLE(file, signature);
	
	file.close();
	
	cout << "balance.dat generated successfully.\n";
	cout << "Money: " << value << endl;
	
	return 0;
}
